cstore_reader.c

```diff
@@ -1287,76 +1354,78 @@ ColumnDefaultValue(TupleConstr *tupleConstraints, Form_pg_attribute attributeFor
 }
 
 
-/* Returns the size of the given file handle. */
-static int64
-FileSize(FILE *file)
+/* ReadFromFile reads the given segment from the relation data file. */
+static StringInfo
+ReadFromFile(Relation relation, uint64 offset, uint32 size)
 {
-	int64 fileSize = 0;
-	int fseekResult = 0;
+	int blockCapacity = CSTORE_PAGE_DATA_SIZE;
+	uint32 blockNumber = offset / blockCapacity;
+	uint32 blockOffset = offset % blockCapacity;
+	uint32 remainingSize = size;
+	uint32 resultOffset = 0;
+	uint32 blockCount = 0;
+	StringInfo resultBuffer = NULL;
 
-	errno = 0;
-	fseekResult = fseeko(file, 0, SEEK_END);
-	if (fseekResult != 0)
+	if (size == 0)
 	{
-		ereport(ERROR, (errcode_for_file_access(),
-						errmsg("could not seek in file: %m")));
+		return resultBuffer;
 	}
 
-	fileSize = ftello(file);
-	if (fileSize == -1)
+	resultBuffer = makeStringInfo();
+	enlargeStringInfo(resultBuffer, size);
+
+	blockCount = RelationGetNumberOfBlocksInFork(relation,DATA_FORKNUM);
+	if (blockNumber > blockCount)
 	{
-		ereport(ERROR, (errcode_for_file_access(),
-						errmsg("could not get position in file: %m")));
+		ereport(ERROR, (errmsg("could not read from file")));
 	}
 
-	return fileSize;
-}
-
+	while (remainingSize > 0)
+	{
+		Buffer buffer = ReadBufferExtended(relation, DATA_FORKNUM, blockNumber,
+										   RBM_NORMAL, NULL);
```

Comment:
almost calls to ReadBufferExtended passes NULL as strategy in the postgres source. 

We can do ```GetAccessStrategy(BAS_BULKREAD)``` for bulk read operation.